File: Day06/06.R
library(tidyverse)

# temp <- readLines("./Day06/example.txt")
temp <- readLines("./inputs/06.txt")

border <- length(temp)
guard_map <- temp |>
  str_split("") |>
  unlist() |>
  matrix(nrow = border, byrow = TRUE)

start <- which(guard_map == "^", arr.ind = TRUE) |> as.numeric()
pos <- start
move <- c(-1, 0)
turn <- function(move) c(move[2], -move[1])

while (TRUE) {
  guard_map[pos[1], pos[2]] <- "X"
  temp <- pos + move

  if (guard_map[temp[1], temp[2]] == "#") move <- turn(move)

  pos <- pos + move
}

sum(guard_map == "X")

reset_guard_map <- function() {
  pos <<- start
  move <<- c(-1, 0)
  guard_map[guard_map == "X"] <<- "."
  guard_map[start[1], start[2]] <<- "^"
}

reset_guard_map()

is_loop <- function(guard_map, O, pos, move) {
  guard_map_copy <- guard_map
  guard_map_copy[O[1], O[2]] <- "#"

  visited_positions <- list()

  tryCatch(
    {
      while (TRUE) {
        current_state <- list(position = pos, direction = move)
        if (any(sapply(visited_positions, function(x) identical(x, current_state)))) {
          return(TRUE)
        }

        visited_positions <- append(visited_positions, list(current_state))

        temp <- pos + move

        if (guard_map_copy[temp[1], temp[2]] == "#") {
          move <- turn(move)
        }

        pos <- pos + move
      }
    },
    error = function(cnd) {
      return(FALSE)
    }
  )
}

# should work, just slow.
library(furrr)
plan(multisession)
temp <- which(guard_map == ".", arr.ind = TRUE)
future_map2_lgl(temp[, 1], temp[, 2], ~ is_loop(guard_map, c(.x, .y), pos, move), .progress = TRUE) |> sum()

File: Day06/06.jl
temp = readlines("./Day06/example.txt")
# temp = readlines("./inputs/06.txt")

border = length(temp)
temp = [collect(line) for line in temp]
guard_map = Matrix{Char}(undef, border, border)

for i in eachindex(temp)
    for j in 1:length(temp[i])
        guard_map[i, j] = temp[i][j]
    end
end

start = findall(x -> x == '^', guard_map)[1]
pos = start
move = CartesianIndex(-1, 0)

function turn(move)
    CartesianIndex(move[2], -move[1])
end

while true
    # show(stdout, "text/plain", guard_map)
    guard_map[pos] = 'X'
    temp_pos = pos + move

    if guard_map[temp_pos] == '#'
        move = turn(move)
    end

    pos = pos + move
    println(pos)
end

length(findall(x -> x == 'X', guard_map))

function reset_guard_map!()
    for i in 1:border
        for j in 1:border
            if guard_map[i, j] == 'X'
                guard_map[i, j] = '.'
            end
        end
    end
    guard_map[start] = '^'

    nothing
end

reset_guard_map!()

function is_loop(guard_map, O)
    pos = start
    move = CartesianIndex(-1, 0)
    guard_map_copy = copy(guard_map)
    guard_map_copy[O] = '#'

    visited_positions = []

    try
        while true
            current_state = (pos, move)
            # println(current_state)
            # show(stdout, "text/plain", guard_map_copy)
            if any(x -> x == current_state, visited_positions)
                return true
            end

            push!(visited_positions, current_state)

            temp = pos + move

            if guard_map_copy[temp] == '#'
                move = turn(move)
            end

            pos = pos + move
        end
    catch e
        return false
    end
end


empty_positions = findall(x -> x == '.', guard_map)

using Base.Threads
results = zeros(Int, length(empty_positions))
@threads for i in eachindex(empty_positions)
    results[i] = is_loop(guard_map, empty_positions[i])
end
sum(results)
